hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -0,0 +1,332 @@
+   * @param favoredNodes - list of nodes preferred. This is a hint to the
+   *                     allocator, whether the favored nodes will be used
+   *                     depends on whether the nodes meets the allocator's
+   *                     requirement.
+   * @param nodesRequired - number of datanodes required.
+   * @param dataSizeRequired - size required for the container.
+   * @param metadataSizeRequired - size required for Ratis metadata.
+   * @return List of datanodes.
+   * @throws SCMException  SCMException
+   */
+  @Override
+  public List<DatanodeDetails> chooseDatanodes(
+      List<DatanodeDetails> excludedNodes, List<DatanodeDetails> favoredNodes,
+      int nodesRequired, long metadataSizeRequired, long dataSizeRequired)
+      throws SCMException {
+    Preconditions.checkArgument(nodesRequired > 0);
+    metrics.incrDatanodeRequestCount(nodesRequired);
+    int datanodeCount = networkTopology.getNumOfLeafNode(NetConstants.ROOT);
+    int excludedNodesCount = excludedNodes == null ? 0 : excludedNodes.size();
+    if (datanodeCount < nodesRequired + excludedNodesCount) {
+      throw new SCMException("No enough datanodes to choose. " +
+          "TotalNode = " + datanodeCount +
+          " RequiredNode = " + nodesRequired +
+          " ExcludedNode = " + excludedNodesCount, null);
+    }
+
+    List<DatanodeDetails> mutableFavoredNodes = new ArrayList<>();
+    if (favoredNodes != null) {
+      // Generate mutableFavoredNodes, only stores valid favoredNodes
+      for (DatanodeDetails datanodeDetails : favoredNodes) {
+        if (isValidNode(datanodeDetails, metadataSizeRequired,
+            dataSizeRequired)) {
+          mutableFavoredNodes.add(datanodeDetails);
+        }
+      }
+      Collections.shuffle(mutableFavoredNodes);
+    }
+    if (excludedNodes != null) {
+      mutableFavoredNodes.removeAll(excludedNodes);
+    }
+
+    // For excluded nodes, we sort their racks at rear
+    List<Node> racks = getAllRacks();
+    if (excludedNodes != null && excludedNodes.size() > 0) {
+      racks = sortRackWithExcludedNodes(racks, excludedNodes);
+    }
+
+    List<Node> toChooseRacks = new LinkedList<>(racks);
+    List<Node> chosenNodes = new ArrayList<>();
+    List<Node> unavailableNodes = new ArrayList<>();
+    Set<Node> skippedRacks = new HashSet<>();
+    if (excludedNodes != null) {
+      unavailableNodes.addAll(excludedNodes);
+    }
+
+    // If the result doesn't change after retryCount, we return with exception
+    int retryCount = 0;
+    while (nodesRequired > 0) {
+      if (retryCount > OUTER_LOOP_MAX_RETRY) {
+        throw new SCMException("No satisfied datanode to meet the" +
+            " excludedNodes and affinityNode constrains.", null);
+      }
+      int chosenListSize = chosenNodes.size();
+
+      // Refill toChooseRacks, we put skippedRacks in front of toChooseRacks
+      // for a even distribution
+      if (toChooseRacks.size() == 0) {
```

Comment:
The only way for toChooseRacks.size() to equal zero, is if we don't have enough racks in the cluster to meet the goal.

Lets say we have 10 racks and EC 3-2, so we need 5 nodes and ideally 5 racks. 6 racks are "bad" for some reason - they have zero nodes or all nodes are full.

On the first pass of the outerloop, we will pick 4 nodes form 4 racks, and skip 6 racks.

On the next pass, we will try the same skipped 6 again, as that is all which is left in `toChooseRacks`.

On the next pass it will be the same, and eventually we will error.

I think we need to refill toChooseRacks on each pass of the outloop, with the skipped racks first. That way:

On the first pass we will get 4 nodes on 4 racks. 6 skipped racks

On the next pass, we will give the skipped racks another try first, but they will all still fail, then we will pick a node from a remaining good rack. We have not met the goal, but we have fallen back as not enough good racks are available.

Reply:
I think that is the current implementation?
First pass: 4 racks chosen, 6 racks added to skipped Racks.
Second pass: 6 racks in skipped racks added in front of toChooseRacks, but the first  6 racks still got no result, so we choose the 5 rack, which will be same as one of the first 4 racks chosen in first pass.

Reply:
We only refill `toChooseRacks` if `toChooseRacks.size() == 0` and we only remove a value from it when we successfully pick a node from a rack. So after pass 1 in my example, there will be 6 nodes left in `toChooseRacks` and the same 6 in skipped, and we will just keep retrying the same skipped list. Here is where we remove from toChooseRacks:

```
if (nodes.size() > 0) {
          Node affinityNode = nodes.get(0);
          Node node = chooseNode(unavailableNodes, affinityNode,
              metadataSizeRequired, dataSizeRequired);
          if (node != null) {
            chosenNodes.add(node);
            mutableFavoredNodes.remove(node);
            unavailableNodes.add(node);
            nodesRequired--;
            if (nodesRequired == 0) {
              break;
            }
          }
        } else {
          // Store the skipped racks to check them first in next outer loop
          skippedRacks.add(rack);
        }
```

Reply:
Sorry. I missed where you always add to `chosenRacksInForLoop`.

In that case, can we avoid adding to `chosenRacksInForLoop`, and also avoid the line `toChooseRacks.removeAll(chosenRacksInForLoop);` and just refill the `toChooseRacks` always on each pass?